### PR TITLE
feat: add z-order optimize

### LINKED ISSUE #1127: Implement Z-order sorting option in Optimize operation
# Description

Would like to be able to z-order tables. We probably need to implement a z-order sorting function in apache/arrow-rs first.

**Use Case**

**Related Issue(s)**

Probably should prioritize #1125 first.

### PR DESCRIPTION
**Pull request by wjones127**
# Description

Implements Z-order in Rust. This is a very basic version that requires loading the whole partition into memory for sorting. In the future, we can implement a DataFusion-based code path that allows sorting with spilling to disk for even larger optimize jobs.

The Z-order function here is based on Arrow's row format. We truncate to take the first 16 bytes of data (padding with zeros at end as necessary). So for variable-width columns like strings, this means that we are only using the first 15 bytes of the string (the first byte is used to differentiate null and empty strings, [see row format docs](https://docs.rs/arrow-row/40.0.0/arrow_row/struct.RowConverter.html#variable-length-bytes-including-strings-encoding)). If a user has a string column where they all share the same prefix, this z-order function won't work well for them. But in many common cases it will work.

We'll also expose this in Python as a follow up.

# Related Issue(s)

- closes #1127

# Documentation

### CODE REVIEW

**Comment by roeap** [comment]
took the liberty of resolving the merge conflicts I caused :)

**Review by roeap — DISMISSED** [review]
Very nice ❤️ - folks are going to be quite excited about this ...

Left a few minor comments, mainly to also better understand the inner workings :).

File: rust/src/operations/optimize.rs [hunk 1]
```diff
@@ -414,6 +439,56 @@ impl MergePlan {
         Ok((partial_actions, partial_metrics))
     }
 
+    /// Creates a stream of batches that are zordered
+    ///
+    /// Currently requires loading all the data into memory.
```
[the remark this answers is not in the document]
Comment (wjones127) [reply rust/src/operations/optimize.rs:444]: Yeah if it's non-partitioned, it's everything. If it's partitioned then only `max_concurrent_tasks` number of partitions are loaded at once.

File: rust/src/operations/optimize.rs [hunk 2]
```diff
@@ -664,3 +789,340 @@ fn build_compaction_plan(
...
+    pub fn sort_chunked_indices(array: Vec<ArrayRef>) -> Result<Vec<(usize, usize)>, ArrowError> {
+        let single_array = arrow_select::concat::concat(
+            array
+                .iter()
+                .map(|arr| arr.as_ref())
+                .collect_vec()
+                .as_slice(),
+        )
+        .unwrap();
+        let indices = arrow::compute::lexsort_to_indices(
```
[the remark this answers is not in the document]
Comment (wjones127) [reply rust/src/operations/optimize.rs:878]: Oh I think you are right! I'll give that a shot.

File: rust/src/operations/optimize.rs [hunk 3]
```diff
@@ -333,6 +360,13 @@ impl MergePlan {
         partition: PartitionTuples,
         files: MergeBin,
         object_store: ObjectStoreRef,
+        read_stream: impl Future<
+                Output = Result<
+                    BoxStream<'static, Result<RecordBatch, ParquetError>>,
+                    DeltaTableError,
+                >,
+            > + Send
+            + 'static,
```
[the remark this answers is not in the document]
Comment (wjones127) [reply rust/src/operations/optimize.rs]: LMK what you think of the type I added.

File: rust/src/operations/optimize.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wjones127) [reply rust/src/operations/optimize.rs]: Added a comment for this.

File: rust/src/operations/optimize.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wjones127) [reply rust/src/operations/optimize.rs]: Nicely simplified :)

**Review by roeap — APPROVED** [review]
Great Work!

**Pull request merged**